### PR TITLE
Add Benchmark Framework for ducktape

### PR DESCRIPTION
**Pull request by k-raina**
What
----
**Key Features:**
- **MetricsCollector**: Real-time performance metrics collection with latency tracking, memory monitoring, and throughput analysis
- **MetricsBounds**: Configurable performance thresholds with automatic validation
- **Enhanced Tests**: All existing ducktape tests now include integrated benchmark metrics
- **Rich Reporting**: Detailed performance reports with P50/P95/P99 latencies, memory usage, and batch efficiency

**Metrics Collected:**
- Throughput: Send/delivery rates (msg/s, MB/s) with realistic bounds (1k+ msg/s)
- Latency: P50/P95/P99 percentiles using Python's `statistics.quantiles()`
- Memory: Peak usage and growth tracking via `psutil`
- Efficiency: Messages per poll, buffer utilization, per-topic/partition breakdowns
- Reliability: Success/error rates with comprehensive validation

**Files Added:**
- `tests/ducktape/benchmark_metrics.py` - Complete benchmark framework

**Files Modified:**  
- `tests/ducktape/test_producer.py` - Enhanced all tests with integrated metrics
- `tests/ducktape/README.md` - Updated documentation

Checklist
------------------
- [x] Contains customer facing changes? Including API/behavior changes
  - No breaking changes - all existing tests enhanced with metrics, not replaced
- [x] Did you add sufficient unit test and/or integration test coverage for this PR?
  - Yes - all existing ducktape tests now include comprehensive metrics validation
  - Validated with 348k+ msg/s throughput and sub-100ms P95 latency

References
----------
- Original TODO: https://github.com/confluentinc/confluent-kafka-python/pull/2021#discussion_r2277930746

Test & Review
------------

```bash
# Run enhanced ducktape tests with integrated benchmarks
./tests/ducktape/run_ducktape_test.py
```

### CODE REVIEW

File: tests/ducktape/benchmark_metrics.py [hunk 1]
```diff
@@ -0,0 +1,394 @@
...
+        try:
+            # Initialize process object once for efficiency
+            if self._process is None:
+                self._process = psutil.Process(os.getpid())
+            
+            return self._process.memory_info().rss / (1024 * 1024)
+        except (psutil.NoSuchProcess, psutil.AccessDenied, psutil.ZombieProcess):
+            # Handle edge cases where process might not exist or be accessible
+            return None
+        except Exception:
```
Comment (MSeal) [inline tests/ducktape/benchmark_metrics.py]: I would not catch generic Exception here and just let it boil up to be remediated

File: tests/ducktape/benchmark_metrics.py [hunk 2]
```diff
@@ -0,0 +1,394 @@
...
+            # Handle any other psutil-related errors
+            return None
+
+
+class MetricsBounds:
```
Comment (MSeal) [inline tests/ducktape/benchmark_metrics.py:250]: Maybe add a TODO: load from config file?

File: tests/ducktape/test_producer.py [hunk 3]
```diff
@@ -61,20 +58,31 @@ def test_basic_produce(self):
...
-
-        # Time-based message production
-        self.logger.info("Producing messages for %.1f seconds to topic %s", test_duration, topic_name)
+                # Calculate actual latency if we have send time
+                msg_key = msg.key().decode('utf-8', errors='replace') if msg.key() else 'unknown'
+                if msg_key in send_times:
+                    latency_ms = (time.time() - send_times[msg_key]) * 1000
+                    del send_times[msg_key]  # Clean up
+                else:
+                    latency_ms = 5.0  # Default latency if timing info not available
```
Comment (MSeal) [inline tests/ducktape/test_producer.py]: maybe better to just set to 0 or None

**Review by MSeal — CHANGES_REQUESTED** [review]
Minor comments. I was debating if we should use something like locust for this.. might be worth switching to down the road but you kind of have to hack it to do any non-RESTful patterns for testing. e.g. https://github.com/SvenskaSpel/locust-plugins/blob/master/examples/kafka_ex.py

**Comment by MSeal** [comment]
Let's touch up small things, get a merge then iterate / change things if we want later. I want to get this into the history so we can build abstractions above for simpler test definitions and swap the implementation details as needed / remove conflicts on future PRs

File: tests/ducktape/benchmark_metrics.py (hunk 2 above)
Comment (k-raina) [reply tests/ducktape/benchmark_metrics.py:250]: Implemented in commit https://github.com/confluentinc/confluent-kafka-python/pull/2030/commits/eb493bb9f4020809518afdca328c67baeb53d931

File: tests/ducktape/benchmark_metrics.py [hunk 4]
```diff
@@ -0,0 +1,414 @@
...
+                partition_summary[partition_key] = {
+                    'messages_sent': metrics['sent'],
+                    'messages_delivered': metrics['delivered'],
+                    'success_rate': metrics['delivered'] / metrics['sent'],
+                    'avg_latency_ms': statistics.mean(metrics['latencies']) if metrics['latencies'] else 0,
+                    'total_bytes': metrics['bytes']
+                }
+
+        return {
+            # Basic metrics
```
Comment (naxinf) [inline tests/ducktape/benchmark_metrics.py:192]: Is the basic vs enhanced classification coming from some other source (e.g. some client benchmarking guides)? The list LGTM but just curious :)

File: tests/ducktape/README.md [hunk 5]
```diff
@@ -1,24 +1,73 @@
...
+- **Per-Topic/Partition Metrics**: Detailed breakdown by topic and partition
+- **Memory Monitoring**: Peak memory usage and growth tracking with psutil
+- **Batch Efficiency**: Messages per poll and buffer utilization analysis
+- **Throughput Validation**: Messages/sec and MB/sec with configurable bounds checking
+- **Comprehensive Reporting**: Detailed performance reports with pass/fail validation
+- **Automatic Bounds Validation**: Performance assertions against configurable thresholds
+
+## Configuration
+
+Performance bounds are loaded from a JSON config file. By default, it loads `benchmark_bounds.json`, but you can override this with the `BENCHMARK_BOUNDS_CONFIG` environment variable:
```
Comment (naxinf) [inline tests/ducktape/README.md:40]: Is default benchmark_bounds.json going to be added in the next PR

File: tests/ducktape/benchmark_metrics.py (hunk 4 above)
Comment (k-raina) [reply tests/ducktape/benchmark_metrics.py:192]: There were no guidelines i refered to, it made sense to me having basic metrics and enhances metrics segregated for ease of reviewer. 

Later on we can further divide these metrics in code/comments as "latency", "throughput", "message delivery" etc.

File: tests/ducktape/README.md (hunk 5 above)
Comment (k-raina) [reply tests/ducktape/README.md:40]: Added bounds in latest commit https://github.com/confluentinc/confluent-kafka-python/pull/2030/commits/d0f4793530c5e30149d336e9743c1050fb5bd6ec

File: tests/ducktape/benchmark_metrics.py (hunk 4 above)
Comment (naxinf) [reply tests/ducktape/benchmark_metrics.py:192]: Make sense!

**Review by fangnx — APPROVED** [review]
